src/main/java/de/tum/in/www1/artemis/config/migration/entries/MigrationEntry20240103_143700.java

```diff
@@ -0,0 +1,411 @@
+     * This is done by cloning the repository from Bitbucket and then creating a bare repository in the local VCS.
+     *
+     * @param exercise      the programming exercise
+     * @param repositoryUrl the repository URL
+     * @return the URL of the repository in the local VCS
+     * @throws GitAPIException    if an error occurs during the git operation
+     * @throws URISyntaxException if the repository URL is invalid
+     */
+    private String cloneRepositoryFromBitbucketAndMoveToLocalVCS(ProgrammingExercise exercise, String repositoryUrl) throws GitAPIException, URISyntaxException {
+        if (localVCService.isEmpty() || bitbucketService.isEmpty() || bitbucketLocalVCMigrationService.isEmpty()) {
+            log.error("Failed to clone repository from Bitbucket: {}", repositoryUrl);
+            return null;
+        }
+        if (repositoryUrl.startsWith(bitbucketLocalVCMigrationService.get().getLocalVCBaseUrl().toString())) {
+            log.info("Repository {} is already in local VC", repositoryUrl);
+            return repositoryUrl;
+        }
+        var repositoryName = uriService.getRepositorySlugFromRepositoryUriString(repositoryUrl);
+        var projectKey = exercise.getProjectKey();
+        Repository oldRepository = gitService.getOrCheckoutRepository(new VcsRepositoryUri(repositoryUrl), true, bitbucketLocalVCMigrationService.get().getDefaultBranch());
+        if (oldRepository == null) {
+            log.error("Failed to clone repository from Bitbucket: {}", repositoryUrl);
+            return null;
+        }
+        localVCService.get().createProjectForExercise(exercise);
+        cloneRepo(projectKey, repositoryName, repositoryUrl);
+        var url = new LocalVCRepositoryUri(projectKey, repositoryName, bitbucketLocalVCMigrationService.get().getLocalVCBaseUrl());
+        return url.toString();
+    }
+
+    /**
+     * Clones the repository from the old origin and creates a bare repository in the local VCS, just as a normal local VC repository would be created
+     * by Artemis.
+     *
+     * @param projectKey     the project key
+     * @param repositorySlug the repository slug
+     * @param oldOrigin      the old origin of the repository
+     */
+    private void cloneRepo(String projectKey, String repositorySlug, String oldOrigin) {
+        LocalVCRepositoryUri localVCRepositoryUri = new LocalVCRepositoryUri(projectKey, repositorySlug, bitbucketLocalVCMigrationService.get().getLocalVCBaseUrl());
+
+        Path remoteDirPath = localVCRepositoryUri.getLocalRepositoryPath(bitbucketLocalVCMigrationService.get().getLocalVCBasePath());
+
+        try {
+            Files.createDirectories(remoteDirPath);
+
+            // Create a bare local repository with JGit.
+            Git git = Git.cloneRepository().setDirectory(remoteDirPath.toFile()).setBare(true).setURI(oldOrigin)
+                    .setBranch(bitbucketLocalVCMigrationService.get().getDefaultBranch()).call();
```

Comment:
Using the default branch here is ok imo (the migration will then convert the old `master` branches to `main`) but then you also need to update the branch saved in the Participation.